src/oscore/oscore_cbor.c

```diff
@@ -0,0 +1,430 @@
+  uint8_t *pt = *buffer;
+  size_t nb = oscore_cbor_put_unsigned(buffer, buf_size, bytes_len);
+  assert(*buf_size >= bytes_len);
+  (*buf_size) -= bytes_len;
+  *pt = (*pt | 0x40);
+  memcpy(*buffer, bytes, bytes_len);
+  (*buffer) += bytes_len;
+  return nb + bytes_len;
+}
+
+size_t
+oscore_cbor_put_map(uint8_t **buffer, size_t *buf_size, uint64_t elements) {
+  uint8_t *pt = *buffer;
+  size_t nb = oscore_cbor_put_unsigned(buffer, buf_size, elements);
+  *pt = (*pt | 0xa0);
+  return nb;
+}
+
+size_t
+oscore_cbor_put_number(uint8_t **buffer, size_t *buf_size, int64_t value) {
+  if (value < 0)
+    return oscore_cbor_put_negative(buffer, buf_size, -value);
+  else
+    return oscore_cbor_put_unsigned(buffer, buf_size, value);
+}
+
+size_t
+oscore_cbor_put_simple_value(uint8_t **buffer,
+                             size_t *buf_size,
+                             uint8_t value) {
+  uint8_t *pt = *buffer;
+  size_t nb = oscore_cbor_put_unsigned(buffer, buf_size, value);
+  *pt = (*pt | 0xe0);
+  return nb;
+}
+
+size_t
+oscore_cbor_put_tag(uint8_t **buffer, size_t *buf_size, uint64_t value) {
+  uint8_t *pt = *buffer;
+  size_t nb = oscore_cbor_put_unsigned(buffer, buf_size, value);
+  *pt = (*pt | 0xc0);
+  return nb;
+}
+
+size_t
+oscore_cbor_put_negative(uint8_t **buffer, size_t *buf_size, int64_t value) {
+  value--;
+  uint8_t *pt = *buffer;
+  size_t nb = oscore_cbor_put_unsigned(buffer, buf_size, value);
+  *pt = (*pt | 0x20);
+  return nb;
+}
+
+static void
+put_b_f(uint8_t **buffer, uint64_t value, uint8_t nr) {
+  uint8_t *pt = *buffer - 1;
+  uint64_t vv = value;
+  for (int q = nr; q > -1; q--) {
+    (*pt--) = (uint8_t)(vv & 0xff);
+    vv = (vv >> 8);
+  }
+}
+
+size_t
+oscore_cbor_put_unsigned(uint8_t **buffer, size_t *buf_size, uint64_t value) {
+  if (value < 0x18) { /* small value half a byte */
+    assert(*buf_size >= 1);
+    (*buf_size)--;
+    (**buffer) = (uint8_t)value;
+    (*buffer)++;
+    return 1;
+  } else if ((value > 0x17) && (value < 0x100)) {
+    /* one byte uint8_t  */
+    assert(*buf_size >= 2);
+    (*buf_size) -= 2;
+    (**buffer) = (0x18);
+    *buffer = (*buffer) + 2;
+    put_b_f(buffer, value, 0);
+    return 2;
+  } else if ((value > 0xff) && (value < 0x10000)) {
+    /* 2 bytes uint16_t     */
+    assert(*buf_size >= 3);
+    (*buf_size) -= 3;
+    (**buffer) = (0x19);
+    *buffer = (*buffer) + 3;
+    put_b_f(buffer, value, 1);
+    return 3;
+  } else if ((value > 0xffff) && (value < 0x100000000)) {
+    /* 4 bytes uint32_t   */
+    assert(*buf_size >= 5);
+    (*buf_size) -= 5;
+    (**buffer) = (0x1a);
+    *buffer = (*buffer) + 5;
+    put_b_f(buffer, value, 3);
+    return 5;
+  } else /*if(value > 0xffffffff)*/ {
+    /* 8 bytes uint64_t  */
+    assert(*buf_size >= 9);
+    (*buf_size) -= 9;
+    (**buffer) = (0x1b);
+    *buffer = (*buffer) + 9;
+    put_b_f(buffer, value, 7);
+    return 9;
+  }
+}
+
+/* temporary routine to read ascii hex dump */
```

Comment:
The comments above and within the code seem to be unrelated to how the function is actually used.

Reply:
Removed comments and commented out code - missed that in the cleanup of the original code.  Renamed function to get_byte and made it static inline - otherwise getting the byte and bumping the pointer gets more complicated in later code.